### PR TITLE
Challenge: Find The Difference Between Two Strings

### PR DESCRIPTION
**Pull request by anwar099**
Find The Difference Between Two Strings
You are given two strings s and t. 

String t is generated by random shuffling string s and then add one more letter at a random position.

Return the letter that was added to t.

 

Example 1:

Input: s = "abcd", t = "abcde"
Output: "e"
Explanation: 'e' is the letter that was added.
Example 2:

Input: s = "", t = "y"
Output: "y"
 

Constraints:

0 <= s.length <= 1000
t.length == s.length + 1
s and t consist of lowercase English letters.

---
name: solution review 
about: A template PR for code review with a checklist
---

# Behavior

## Files

- [ ] The file name describes the function's behavior
- [ ] There is a module docstring in the function file
- [ ] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [ ] There is a module docstring in the tests file

### Unit Tests

- [ ] The test class has a helpful name in PascalCase
- [ ] The test class has a docstring
- Every unit test has
  - [ ] A helpful name
  - [ ] A clear docstring
  - [ ] Only one assertion
  - [ ] There is no logic in the unit test
- [ ] All tests pass
- [ ] There are tests for defensive assertions
- [ ] There are tests for boundary cases

### Function Docstring

- [ ] The function's behavior is described
- The function's arguments are described:
  - [ ] Type
  - [ ] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [ ] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [ ] Include 3 or more (passing!) doctests

### The Function

- [ ] The function's name describes it's behavior
- [ ] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [ ] The function has correct type annotations
- [ ] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [ ] Variable names help to understand the strategy
- [ ] Any comments are clear and describe the strategy
- [ ] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in any docstrings or tests
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [ ] The code passes the formatting checks
- [ ] The code passes all Ruff linting checks
- [ ] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [ ] Variables are named with snake_case
- [ ] Variable names are clear and helpful
- [ ] The code follows the strategy as simply as possible
- [ ] The implementation is as simple as possible given the strategy
- [ ] There are no commented lines of code
- [ ] There are no `print` statements anywhere
- [ ] The code includes defensive assertions
- [ ] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/Dif_Tw_Str.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+def find_the_difference(s: str, t: str) -> str:
```
Comment (MadiMalik) [inline solutions/Dif_Tw_Str.py:1]: Function Name: The function name, find_the_difference, describes its behavior, But does not match the file name

File: solutions/Dif_Tw_Str.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+    Examples:
+        >>> find_the_difference("abcd", "abcde")
+        'e'
+        >>> find_the_difference("", "y")
+        'y'
+    """
+    result = 0
+    # XOR all characters in both strings
+    for char in s + t:
```
Comment (MadiMalik) [inline solutions/Dif_Tw_Str.py:20]: Variable names like results are clear given the XOR-based strategy.
The function’s single loop over s + t with XOR is efficient and well-documented.

File: solutions/tests/test_find_the_difference.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
+import unittest
+
+from solutions.Dif_Tw_Str import find_the_difference
+
+
+class TestFindTheDifference(unittest.TestCase):
```
Comment (MadiMalik) [inline solutions/tests/test_find_the_difference.py:6]: Test class name is a helpful and descriptive written in PascalCase

**Comment by MadiMalik** [comment]
Hey Anwar,

As I mentioned before, your function name describes its behavior, but it does not match the file name. The function file has a clear module-level docstring explaining the purpose of the file, which is good.

However, the function lacks defensive assertions. No specific assertions are added or documented. Please include assertion checks to validate assumptions about the input and document these checks explicitly in the docstring under a Raises section.

Also, make sure to test for cases where s and t contain unusual input, such as strings with only whitespace or mixed case sensitivity.

Let me know if you need help with this!
